### PR TITLE
✨ Pass consent to amp-brightcove iframe src

### PR DESCRIPTION
**Pull request by mister-ben**
Where amp-consent and amp-brightcove are used together, adds the consent state, consent string and consent data to the player iframe, so that the Brightcove player advertising configuration can be set accordingly.

### CODE REVIEW

**Comment by mister-ben** [comment]
The failing tests are unrealted to these changes.

**Comment by rcebulko** [comment]
@ampproject/wg-components bumping this review for visibility

File: extensions/amp-brightcove/0.1/amp-brightcove.js [hunk 1]
```diff
@@ -81,8 +86,8 @@ class AmpBrightcove extends AMP.BaseElement {
     /** @private {?Function} */
     this.unlistenMessage_ = null;
 
-    /**@private {?string} */
-    this.playerId_ = null;
+    /**@private {string} */
+    this.playerId_ = '';
```
Comment (micajuine-ho) [inline extensions/amp-brightcove/0.1/amp-brightcove.js:90]: nit: undo this change, unless it's necessary?

File: extensions/amp-brightcove/0.1/amp-brightcove.js [hunk 2]
```diff
@@ -123,17 +128,32 @@ class AmpBrightcove extends AMP.BaseElement {
...
+      : Promise.resolve(null);
+  }
 
-    this.iframe_ = iframe;
+  /**
+   * @override
+   * @return {!Promise}
+   */
+  layoutCallback() {
+    this.getIframeSrc_().then((src) => {
```
Comment (micajuine-ho) [inline extensions/amp-brightcove/0.1/amp-brightcove.js:146]: `return this.getIframeSrc_()....`

File: extensions/amp-brightcove/0.1/amp-brightcove.js (hunk 1 above)
Comment (micajuine-ho) [inline extensions/amp-brightcove/0.1/amp-brightcove.js:90]: nit: revert unless necessary

File: extensions/amp-brightcove/0.1/amp-brightcove.js [hunk 3]
```diff
@@ -123,17 +128,32 @@ class AmpBrightcove extends AMP.BaseElement {
     ));
   }
 
-  /** @override */
-  layoutCallback() {
-    const iframe = createFrameFor(this, this.getIframeSrc_());
+  /**
+   * @return {!Promise}
+   */
+  getConsentState() {
```
Comment (micajuine-ho) [inline extensions/amp-brightcove/0.1/amp-brightcove.js:134]: nit: getConsentState() => getConsentState_()

File: extensions/amp-brightcove/0.1/amp-brightcove.js [hunk 4]
```diff
@@ -259,47 +279,94 @@ class AmpBrightcove extends AMP.BaseElement {
...
+      : Promise.resolve();
+    const sharedDataPromise = consentPolicyId
+      ? getConsentPolicySharedData(this.element, consentPolicyId)
+      : Promise.resolve();
+
+    return Promise.all([
+      consentPromise,
+      sharedDataPromise,
+      consentStringPromise,
+    ]).then((consents) => {
```
Comment (micajuine-ho) [inline extensions/amp-brightcove/0.1/amp-brightcove.js:316]: Let's move this all to `getConsentState_()` (and rename it to something like `getConsents_()`), then have that return the array of consents, so that too much work doesn't get jammed in this method.

**Review by micajuine-ho — CHANGES_REQUESTED** [review]

File: extensions/amp-brightcove/0.1/amp-brightcove.js (hunk 4 above)
Comment (mister-ben) [reply extensions/amp-brightcove/0.1/amp-brightcove.js]: thanks for the review, moved in the latest commit

File: extensions/amp-brightcove/0.1/amp-brightcove.js (hunk 2 above)
Comment (micajuine-ho) [reply extensions/amp-brightcove/0.1/amp-brightcove.js:153]: @mister-ben If I'm not mistaken, since layoutCallback returns a promise, I think this whole promise should be returned?

i.e. 
```
layoutCallback() {
  return this.getIframeSrc_().then(....)
}
```

Comment (mister-ben) [reply extensions/amp-brightcove/0.1/amp-brightcove.js:153]: the tests time out with that change

Comment (micajuine-ho) [reply extensions/amp-brightcove/0.1/amp-brightcove.js:153]: Existing tests or new ones?

Comment (mister-ben) [reply extensions/amp-brightcove/0.1/amp-brightcove.js:153]: Existing tests. All of the tests expect the one that intentionally fails to get as far as this point.

File: extensions/amp-brightcove/0.1/test/test-amp-brightcove.js [hunk 5]
```diff
@@ -133,14 +135,27 @@ describes.realWin(
             '/index.html?videoId=ref:amp-test-video&playsinline=true'
         );
 
+        const newSrc = new Promise((resolve) => {
```
Comment (micajuine-ho) [inline extensions/amp-brightcove/0.1/test/test-amp-brightcove.js:138]: Since you changed `layoutCallback()` you're going to have to add this to `beforeEach()` for the whole test file. That's probably why the existing tests are failing when you had `return getIframeSrc_()....`

**Review by micajuine-ho — CHANGES_REQUESTED** [review]

File: extensions/amp-brightcove/0.1/amp-brightcove.js [hunk 6]
```diff
@@ -317,7 +379,9 @@ class AmpBrightcove extends AMP.BaseElement {
       videoId !== undefined
     ) {
       if (this.iframe_) {
-        this.iframe_.src = this.getIframeSrc_();
+        this.getIframeSrc_().then((src) => {
+          this.iframe_.src = src;
+        });
```
Comment (micajuine-ho) [inline extensions/amp-brightcove/0.1/amp-brightcove.js:384]: This won't work--we can't have `mutatedAttributeCallback` be async. 

I'm thinking we can do something like this:
```
this.consentState = null;
this.sharedData = null;
this.consentString = null;

// Wait for consents and save them
getConsents() {
 ...
 Promise.all(consentsArray).then(arr => {
   this.consentState = arr[0];
   this.sharedData = arr[1];
   this.consentString = arr[2];
 });
}

// Should only be called after our consents have been stored
getIframeSrc_() {
 ...
 if (this.consentState) {
        src = addParamToUrl(
          src,
          'ampInitialConsentState',
          this.consentState
        );
      }
 ...
}

// Get consents then build the src
layoutCallback() {
   return this.getConsents().then(() => {
       const iframe = createFrameFor(this, this.getIframeSrc_());
       return loadPromise....
   }
}

// Change the src only once the consents have been save and the iframe has been created
mutatedAttributesCallback(mutations) {
 ...
 if (this.iframe_) {
  this.iframe_.src = this.getIframeSrc_();
 }
}
```

The idea here is to make getIframeSrc_ synchronous. This should be safe since, we only change the src within `mutatedAttributesCallback` once the iframe_ has been created in `layoutCallback`, at which point our consents will be stored.

**Review by micajuine-ho — CHANGES_REQUESTED** [review]

File: extensions/amp-brightcove/0.1/amp-brightcove.js (hunk 6 above)
Comment (mister-ben) [reply extensions/amp-brightcove/0.1/amp-brightcove.js]: ok, refactored to work this way

File: extensions/amp-brightcove/0.1/test/test-amp-brightcove.js [hunk 7]
```diff
@@ -73,8 +75,8 @@ describes.realWin(
       });
     }
 
-    it('renders', () => {
-      return getBrightcove({
+    it('renders', async () => {
+      getBrightcove({
```
Comment (micajuine-ho) [inline extensions/amp-brightcove/0.1/test/test-amp-brightcove.js:79]: If you change these test to `async` then you need to use async/await for these methods (`const bc = await getBrightcove();`).

However, I recommend just keeping the tests synchronous and returning the promise.

The reason why the tests were failing with the original setup is because we've added a new promise within `layoutCallback()` that we have to wait for.

Within the `getBrightcove()` helper method in the test, we need to wait for this promise to resolve by using `await macroTask()`:

```
      await element.signals().whenSignal(CommonSignals.LOAD_START);
      // wait for consents 
      await macroTask();

      try {
        fakePostMessage(element, {event: 'ready'});
      } catch (_) {
``` 

With this change, you won't need to change the existing tests.

**Review by micajuine-ho — CHANGES_REQUESTED** [review]

File: extensions/amp-brightcove/0.1/test/test-amp-brightcove.js (hunk 7 above)
Comment (mister-ben) [reply extensions/amp-brightcove/0.1/test/test-amp-brightcove.js]: thanks, updated

File: extensions/amp-brightcove/0.1/test/test-amp-brightcove.js [hunk 8]
```diff
@@ -74,7 +79,7 @@ describes.realWin(
     }
 
     it('renders', () => {
-      return getBrightcove({
+      getBrightcove({
```
Comment (micajuine-ho) [inline extensions/amp-brightcove/0.1/test/test-amp-brightcove.js:82]: This still needs to be `return getBrightcove(...`  for all these tests

**Review by micajuine-ho — CHANGES_REQUESTED** [review]

File: extensions/amp-brightcove/0.1/test/test-amp-brightcove.js [hunk 9]
```diff
@@ -133,14 +138,27 @@ describes.realWin(
...
+            observer.disconnect();
+          };
+
+          const observer = new MutationObserver(cb);
+          observer.observe(iframe, {
+            attributes: true,
+            attributesFilter: ['src'],
+          });
+        });
+
```
Comment (micajuine-ho) [inline extensions/amp-brightcove/0.1/test/test-amp-brightcove.js:153]: Is this necessary? I don't think your changes should have affected this.

**Review by micajuine-ho — CHANGES_REQUESTED** [review]

File: extensions/amp-brightcove/0.1/test/test-amp-brightcove.js (hunk 9 above)
Comment (mister-ben) [reply extensions/amp-brightcove/0.1/test/test-amp-brightcove.js]: right, it isn't now getIframeSrc_() is synchronous again

**Review by micajuine-ho — APPROVED** [review]
